tool/cargo/Syncer.kt

```diff
@@ -0,0 +1,419 @@
+                                containsGeneratedSources = props.getProperty(CONTAINS_GENERATED_SOURCES).toBoolean(),
+                                tests = mutableListOf(),
+                                buildScripts = mutableListOf(),
+                        )
+                    } catch (e: Exception) {
+                        throw IllegalStateException("Failed to parse Manifest Sync properties file at $path", e)
+                    }
+                }
+
+                private fun extractDependencyEntries(props: Properties): Map<String, String> {
+                    return props.entries
+                            .map { it.key.toString() to it.value.toString() }
+                            .filter { it.first.startsWith("$DEPS_PREFIX.") }
+                            .toMap()
+                }
+
+                private fun parseDependencies(raw: Map<String, String>): Collection<Dependency> {
+                    return raw.map { Dependency.of(it.key, it.value) }
+                }
+            }
+
+            private object Keys {
+                const val BUILD_DEPS = "build.deps"
+                const val DEPS_PREFIX = "deps"
+                const val EDITION = "edition"
+                const val ENTRY_POINT_PATH = "entry.point.path"
+                const val FEATURES = "features"
+                const val LABEL = "label"
+                const val NAME = "name"
+                const val PATH = "path"
+                const val ROOT_PATH = "root.path"
+                const val TYPE = "type"
+                const val CONTAINS_GENERATED_SOURCES = "contains.generated.sources"
+                const val VERSION = "version"
+            }
+        }
+
+        private object Paths {
+            const val BAZEL_BIN = "bazel-bin"
+            const val CARGO_TOML = "Cargo.toml"
+            const val EXTERNAL = "external"
+            const val EXTERNAL_PLACEHOLDER = "{external}"
+            const val MANIFEST_SYNC_PROPERTIES_SUFFIX = ".cargo-sync.properties"
+        }
+
+        companion object {
+            const val GENERATED_FILE_NOTICE =
+"""
+# Generated by Vaticle Cargo sync tool.
+# Do not commit or modify this file.
+
+"""
+        }
+    }
+
+    private object ShellArgs {
+        const val ASPECTS = "--aspects"
+        const val BAZEL = "bazel"
+        const val BAZEL_BIN = "bazel-bin"
+        const val BUILD = "build"
+        const val INFO = "info"
+        const val OUTPUT_BASE = "output_base"
+        const val OUTPUT_GROUPS_RUST_CARGO_SYNC_PROPERTIES = "--output_groups=rust-cargo-sync-properties"
+        const val QUERY = "query"
+        const val RUST_CARGO_SYNC_PROPERTIES_ASPECT = "@vaticle_dependencies//tool/cargo:sync_properties_aspect.bzl%rust_cargo_sync_properties_aspect"
+        const val RUST_TARGETS_DEPS_QUERY = "kind(rust_*, deps(kind(rust_*, //...)))" // deps of RUST_TARGETS_QUERY
```

Comment:
Note: we replaced `kind(rust_*, deps(//...))` with just searching the rust dependencies of the top-level rust dependencies, because calling `deps` seems to induce an unneeded build phase on Bazel (that happens to break compilation of RocksDB on Windows). In any case, this is not required, nor desirable in the top-level repository.

This could be simplified to: in each workspace, find all `rust_*` targets, and recursively find directly descendant `rust_*` targets. However, this implementation is more far more complex to implement. If we did implement it though, it works because of the idea that a Cargo project can only depend on Cargo projects recursively without breaks - we can't create a valid cargo project from something that looks like `rust_library -> genrule -> rust_library`.